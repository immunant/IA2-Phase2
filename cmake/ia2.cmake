if (LIBIA2_AARCH64)
    set(UBSAN_FLAG "")
else()
    set(UBSAN_FLAG "-fsanitize=undefined")
endif()
# Creates a compartmentalized IA2 target
#
# add_ia2_compartment(<name> <EXECUTABLE|LIBRARY> PKEY <n> SOURCES <src>...
#                     [INCLUDE_DIRECTORIES <dir>...] [LIBRARIES <lib>...]
#                     [ENABLE_UBSAN])
#
# Creates either an IA2 executable or shared library called <name> with the given pkey <n>.
#
# <name> - Target name.
# PKEY (required) - The pkey for the compartment of this binary.
# SOURCES (required) - The set of source files.
# INCLUDE_DIRECTORIES (optional) - include directories used to build the executable. If
#       a relative path this will be interpreted as relative to the current source
#       directory.
# LIBRARIES (optional) - libraries to link against.
# ENABLE_UBSAN - If present, enable UBSAN for this executable.
function(add_ia2_compartment NAME TYPE)
  # Parse options
  set(options ENABLE_UBSAN)
  set(oneValueArgs PKEY)
  set(multiValueArgs LIBRARIES SOURCES INCLUDE_DIRECTORIES)
  cmake_parse_arguments(ARG "${options}" "${oneValueArgs}"
    "${multiValueArgs}" ${ARGN})

  set(VALID_TYPES EXECUTABLE LIBRARY)
  if(NOT "${TYPE}" IN_LIST VALID_TYPES)
    message(FATAL_ERROR "${TYPE} is not a valid type, expected one of ${VALID_TYPES}")
  endif()

  if("${TYPE}" STREQUAL "EXECUTABLE")
    add_executable(${NAME})
  elseif("${ARG_PKEY}" GREATER "0")
    set(UNPADDED_LIB ${NAME}_unpadded)
    add_library(${UNPADDED_LIB} SHARED)
    pad_tls_library(${UNPADDED_LIB} ${NAME})
    set(NAME ${UNPADDED_LIB})
  else()
    add_library(${NAME} SHARED)
  endif()

  # The x86 version is missing a dependency here and libs rely on include path overlap to pick up
  # criterion header. I want to keep spoofed criterion static for simplicity so we add the -I
  # directly here.
  if (LIBIA2_AARCH64)
    target_include_directories(${NAME} PRIVATE
        ${CMAKE_SOURCE_DIR}/misc/spoofed_criterion/include)
  endif()
  target_compile_definitions(${NAME} PRIVATE
    IA2_ENABLE=1
    PKEY=${ARG_PKEY}
  )
  set_target_properties(${NAME} PROPERTIES PKEY ${ARG_PKEY})
  target_compile_options(${NAME} PRIVATE
    "-Werror=incompatible-pointer-types"
    "-fPIC"
  )

  set_property(TARGET ${NAME} PROPERTY ORIGINAL_SOURCES ${ARG_SOURCES})
  set_property(TARGET ${NAME} PROPERTY ORIGINAL_INCLUDE_DIRECTORIES ${ARG_INCLUDE_DIRECTORIES})

  if (ARG_ENABLE_UBSAN)
    # UBSAN requires passing this as both a compiler and linker flag
    target_compile_options(${NAME} PRIVATE ${UBSAN_FLAG})
    target_link_options(${NAME} PRIVATE ${UBSAN_FLAG})
  endif()

  target_link_libraries(${NAME} PRIVATE dl libia2 partition-alloc)
  target_link_options(${NAME} PRIVATE "-Wl,--export-dynamic")

  target_link_libraries(${NAME} PRIVATE ${ARG_LIBRARIES})

  create_compile_commands(${NAME} ${TYPE}
    PKEY ${ARG_PKEY}
    SOURCES ${ARG_SOURCES}
    INCLUDE_DIRECTORIES ${ARG_INCLUDE_DIRECTORIES}
  )

  if("${TYPE}" STREQUAL "EXECUTABLE")
    # Create and link call gates
    add_ia2_call_gates(${NAME} LIBRARIES ${ARG_LIBRARIES})
  endif()
endfunction()

# Prepend <basedir> to given paths, if relative.
#
# relative_to_absolute(<output_var> <basedir> <path>...)
function(relative_to_absolute OUTPUT BASEDIR)
  foreach(dir ${ARGN})
    if("${dir}" MATCHES "^/")
      # absolute path, so we want to leave it alone
      list(APPEND paths "${dir}")
    else()
      list(APPEND paths "${BASEDIR}/${dir}")
    endif()
  endforeach()
  set(${OUTPUT} ${paths} PARENT_SCOPE)
endfunction()

function(pad_tls_library INPUT OUTPUT)
  # Add command and target to generate the padded, sonamed library
  add_custom_command(
    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/lib${OUTPUT}.so
    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:${INPUT}> ${CMAKE_CURRENT_BINARY_DIR}/lib${OUTPUT}.so
    COMMAND ${CMAKE_BINARY_DIR}/tools/pad-tls/pad-tls --allow-no-tls ${CMAKE_CURRENT_BINARY_DIR}/lib${OUTPUT}.so
    DEPENDS ${PAD_TLS_SRCS} tools pad-tls $<TARGET_FILE:${INPUT}>
    COMMENT "Padding TLS segment of wrapped library"
  )
  add_custom_target(${OUTPUT}-padding DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/lib${OUTPUT}.so")
  add_library(${OUTPUT} SHARED IMPORTED GLOBAL)
  set_property(TARGET ${OUTPUT} PROPERTY IMPORTED_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/lib${OUTPUT}.so")
  add_dependencies(${OUTPUT} ${OUTPUT}-padding)

  set_target_properties(${INPUT} PROPERTIES
    LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/unpadded"
    OUTPUT_NAME "${OUTPUT}"
  )
  target_link_options(${OUTPUT} INTERFACE $<TARGET_PROPERTY:${INPUT},INTERFACE_LINK_OPTIONS>)
endfunction()

# Create a fake target that builds the given sources
#
# We use this target to provide the input compile_commands.json for the source
# rewriter.
function(create_compile_commands NAME TYPE)
  # Parse options
  set(options "")
  set(oneValueArgs PKEY)
  set(multiValueArgs SOURCES INCLUDE_DIRECTORIES)
  cmake_parse_arguments(ARG "${options}" "${oneValueArgs}"
    "${multiValueArgs}" ${ARGN})

  relative_to_absolute(SOURCES ${CMAKE_CURRENT_SOURCE_DIR} ${ARG_SOURCES})
  relative_to_absolute(INCLUDE_DIRECTORIES ${CMAKE_CURRENT_SOURCE_DIR} ${ARG_INCLUDE_DIRECTORIES})

  # Create a temporary target to get a compile_commands.json. This target will
  # never be built.
  set(COMPILE_COMMAND_TARGET ${NAME}_compile_commands)
  set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
  if("${TYPE}" STREQUAL "EXECUTABLE")
    add_executable(${COMPILE_COMMAND_TARGET} EXCLUDE_FROM_ALL ${SOURCES})
  else()
    add_library(${COMPILE_COMMAND_TARGET} EXCLUDE_FROM_ALL SHARED ${SOURCES})
  endif()
  target_compile_definitions(${COMPILE_COMMAND_TARGET} PRIVATE
    IA2_ENABLE=0
    PKEY=${ARG_PKEY}
  )
  # Copy target properties from the real target. We might need to add more properties.
  target_link_libraries(${COMPILE_COMMAND_TARGET} PRIVATE $<TARGET_PROPERTY:${NAME},LINK_LIBRARIES>)
  target_include_directories(${COMPILE_COMMAND_TARGET} PRIVATE ${INCLUDE_DIRECTORIES})
  if (LIBIA2_AARCH64)
    target_include_directories(${COMPILE_COMMAND_TARGET} PRIVATE
        ${CMAKE_SOURCE_DIR}/misc/spoofed_criterion/include)
  endif()
  set(CMAKE_EXPORT_COMPILE_COMMANDS OFF)
endfunction()

# We need to get the system headers specifically from clang
find_program(CLANG_EXE
  clang
  PATHS /usr/bin/clang
  DOC "Path to clang compiler")

if(NOT CLANG_EXE)
  message(FATAL_ERROR "Could not find Clang, please set CLANG_EXE manually")
endif()
message(STATUS "Found Clang executable: ${CLANG_EXE}")

# Get system header directories for header rewriter as it needs them explicitly passed
execute_process(COMMAND ${CLANG_EXE} -print-file-name=include
  OUTPUT_VARIABLE CLANG_HEADERS_INCLUDE
  OUTPUT_STRIP_TRAILING_WHITESPACE)
message(STATUS "Found Clang headers: ${CLANG_HEADERS_INCLUDE}")

execute_process(COMMAND ${CLANG_EXE} -print-file-name=include-fixed
  OUTPUT_VARIABLE CLANG_HEADERS_INCLUDE_FIXED
  OUTPUT_STRIP_TRAILING_WHITESPACE)
message(STATUS "Found Clang fixed headers: ${CLANG_HEADERS_INCLUDE_FIXED}")

file(GLOB REWRITER_SRCS ${CMAKE_SOURCE_DIR}/tools/rewriter/*.cpp)
file(GLOB PAD_TLS_SRCS ${CMAKE_SOURCE_DIR}/tools/pad-tls/*.c)
# This cannot be in the tools directory CMakeLists.txt because the target is for
# the top-level CMake project
add_custom_target(rewriter
    COMMAND ${CMAKE_COMMAND} --build . -t ia2-rewriter
    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/tools
    # tools dependency is for the CMake config step
    DEPENDS tools ${REWRITER_SRCS})

add_custom_target(pad-tls
    COMMAND ${CMAKE_COMMAND} --build . -t pad-tls
    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/tools
    DEPENDS tools ${PAD_TLS_SRCS})

# Create call gates for a target
#
# Creates call gates for executable target <name> and all of its
# compartmentalized dependencies. As part of this process we rewrite all given
# input sources for these targets and add the rewritten sources as the new
# sources for the targets (and similarly for include directories to pick up
# rewritten headers).
#
# This depends on our custom target properties PKEY, ORIGINAL_SOURCES, and
# ORIGINAL_INCLUDE_DIRECTORIES being set for any compartmentalized targets.
function(add_ia2_call_gates NAME)
  # Parse options
  set(options "")
  set(oneValueArgs "")
  set(multiValueArgs LIBRARIES EXTRA_REWRITER_ARGS)
  cmake_parse_arguments(ARG "${options}" "${oneValueArgs}"
    "${multiValueArgs}" ${ARGN})

  get_target_property(PKEY ${NAME} PKEY)

  set(CALL_GATE_TARGET ${NAME}_call_gates)
  set(CALL_GATE_SRC ${CMAKE_CURRENT_BINARY_DIR}/${NAME}_call_gates.c)
  set(CALL_GATE_HDR ${CMAKE_CURRENT_BINARY_DIR}/${NAME}_call_gates.h)

  foreach(target ${NAME} ${ARG_LIBRARIES})
    set(target_pkey_set FALSE)
    if(TARGET ${target})
      get_property(target_pkey_set TARGET ${target} PROPERTY PKEY SET)
    endif()
    if(NOT ${target_pkey_set})
      if(TARGET ${target}_unpadded)
        get_property(target_pkey_set TARGET ${target}_unpadded PROPERTY PKEY SET)
      endif()
      if(${target_pkey_set})
        set(target ${target}_unpadded)
      endif()
    endif()

    set(REWRITER_OUTPUT_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/${NAME}_call_gates")

    if(${target_pkey_set})
      get_target_property(target_source_dir ${target} SOURCE_DIR)
      get_target_property(target_binary_dir ${target} BINARY_DIR)

      get_target_property(target_pkey ${target} PKEY)
      set(target_ld_args_file "${REWRITER_OUTPUT_PREFIX}_${target_pkey}.ld")
      target_link_options(${target} PRIVATE "-Wl,@${target_ld_args_file}")
      list(APPEND LD_ARGS_FILES "${target_ld_args_file}")

      set(target_objcopy_args_file "${REWRITER_OUTPUT_PREFIX}_${target_pkey}.objcopy")
      set(OBJCOPY_GLUE ${CMAKE_OBJCOPY} "--redefine-syms=${target_objcopy_args_file}")
      set(OBJCOPY_CMD ${OBJCOPY_GLUE} $<JOIN:$<TARGET_OBJECTS:${target}>, \\\; && ${OBJCOPY_GLUE} >)
      add_custom_command(TARGET ${target} PRE_LINK
                         COMMAND "${OBJCOPY_CMD}"
                         DEPENDS ${target_objcopy_args_file}
                         VERBATIM
                         COMMAND_EXPAND_LISTS)
      list(APPEND OBJCOPY_ARGS_FILES "${target_objcopy_args_file}")

      get_target_property(target_srcs ${target} ORIGINAL_SOURCES)
      relative_to_absolute(target_original_srcs ${target_source_dir} ${target_srcs})
      list(APPEND SOURCES ${target_original_srcs})
      relative_to_absolute(target_rewritten_srcs ${target_binary_dir} ${target_srcs})
      list(APPEND REWRITTEN_SOURCES ${target_rewritten_srcs})
      target_sources(${target} PRIVATE ${target_rewritten_srcs})

      get_target_property(target_include_dirs ${target} ORIGINAL_INCLUDE_DIRECTORIES)
      if(target_include_dirs)
        relative_to_absolute(target_original_include_dirs ${target_source_dir} ${target_include_dirs})
        list(APPEND INCLUDE_DIRECTORIES ${target_original_include_dirs})
        relative_to_absolute(target_rewritten_include_dirs ${target_binary_dir} ${target_include_dirs})
        list(APPEND REWRITTEN_INCLUDE_DIRECTORIES ${target_rewritten_include_dirs})
        if("${target_pkey}" GREATER "0")
          target_include_directories(${target} PRIVATE ${target_rewritten_include_dirs})
        else()
          target_include_directories(${target} PRIVATE ${target_original_include_dirs})
        endif()
      endif()

      # FIXME: This shouldn't be necessary but it seems aarch64-gcc < v13 might
      # default to --as-needed so this is needed to fix some runtime ld.so lookup
      # error
      if (LIBIA2_AARCH64)
          target_link_options(${target} PRIVATE "-Wl,--no-as-needed")
      endif()
      target_link_libraries(${target} PRIVATE ${CALL_GATE_TARGET})

      if("${target_pkey}" GREATER "0")
        # We set this privately on the executable so it doesn't get transitively
        # picked up by the compile command target.
        target_compile_options(${target} PRIVATE "-include${CALL_GATE_HDR}")
      endif()
    endif()
  endforeach()

  if (LIBIA2_AARCH64)
      set(ARCH_FLAG "--arch=aarch64")
  else()
      set(SYSROOT_FLAG
        --extra-arg=-isystem "--extra-arg=${CLANG_HEADERS_INCLUDE}"
        --extra-arg=-isystem "--extra-arg=${CLANG_HEADERS_INCLUDE_FIXED}")
  endif()
  add_custom_command(
    OUTPUT ${CALL_GATE_SRC} ${CALL_GATE_HDR}
           ${LD_ARGS_FILES} ${OBJCOPY_ARGS_FILES} ${REWRITTEN_SOURCES}
    COMMAND ${CMAKE_BINARY_DIR}/tools/rewriter/ia2-rewriter
        --output-prefix=${REWRITER_OUTPUT_PREFIX}
        --root-directory=${CMAKE_CURRENT_SOURCE_DIR}
        --output-directory=${CMAKE_CURRENT_BINARY_DIR}
        ${ARCH_FLAG}
        # Set the build path so the rewriter can find the compile_commands JSON
        -p=${CMAKE_BINARY_DIR}
        ${SYSROOT_FLAG}
        ${ARG_EXTRA_REWRITER_ARGS}
        ${SOURCES}
    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
    # dependencies on custom targets (i.e. the rewriter) do not re-run this
    # command so we need to add a dependency on the rewriter's sources. We still
    # need the dependency on the custom rewriter target to make sure it gets
    # built the first time.
    DEPENDS ${SOURCES} ${REWRITER_SRCS} rewriter
    VERBATIM
  )

  add_custom_target(${NAME}-rewrite
    DEPENDS ${CALL_GATE_SRC} ${CALL_GATE_HDR}
            ${LD_ARGS_FILES} ${OBJCOPY_ARGS_FILES} ${REWRITTEN_SOURCES})

  add_library(${CALL_GATE_TARGET} SHARED ${CALL_GATE_SRC})
  add_dependencies(${CALL_GATE_TARGET} ${NAME}-rewrite)
  # This is needed to enable LIBIA2_DEBUG=1
  target_compile_definitions(${CALL_GATE_TARGET} PRIVATE IA2_ENABLE=1)
  if(LIBIA2_DEBUG)
    target_compile_definitions(${CALL_GATE_TARGET} PRIVATE LIBIA2_DEBUG=1 IA2_ENABLE=1)
  endif()
  target_compile_definitions(${CALL_GATE_TARGET} PRIVATE _GNU_SOURCE)
  target_link_libraries(${CALL_GATE_TARGET} PUBLIC libia2)
endfunction()
