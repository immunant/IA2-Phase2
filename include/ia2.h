#pragma once
#include <link.h>
#include <sys/mman.h>
#include <stdio.h>

// This typedef is required for cbindgen
typedef struct dl_phdr_info dl_phdr_info;

#include "pkey_init.h"

// Attribute for variables that can be accessed from any untrusted compartments.
#define IA2_SHARED_DATA __attribute__((section("ia2_shared_data")))

#ifdef IA2_WRAPPER
#define IA2_WRAP_FUNCTION(name)
#else
#define IA2_WRAP_FUNCTION(name) __asm__(".symver " #name ",__ia2_" #name "@IA2")
#endif

#define XSTR(s) STR(s)
#define STR(s) #s
// TODO: Incorporate __FILE_NAME__ and __COUNTER__ if possible to make this more
// flexible.
#define UNIQUE_STR(s) s "_line_" XSTR(__LINE__)

// On linux protection key 0 is the default for anything not covered by
// pkey_mprotect so untrusted compartments have the lower 2 bits of PKRU cleared
#define PKRU_UNTRUSTED 0xFFFFFFFC

// Suffix for GATE_ macro for untrusted gate code
#define NO_PKEY UNTRUSTED

// Set PKRU to untrusted using rax, r10 and r11 as scratch registers.
#define WRPKRU(pkey)                        \
    "movq %rcx, %r10\n"                     \
    "movq %rdx, %r11\n"                     \
    "movl $" XSTR(pkey) ", %eax\n"           \
    "xorl %ecx, %ecx\n"                     \
    "xorl %edx, %edx\n"                     \
    "wrpkru\n"                              \
    "movq %r10, %rcx\n"                     \
    "movq %r11, %rdx\n"

#ifdef LIBIA2_INSECURE
#define GATE(x)
#define DISABLE_PKRU
#else
#define GATE(pkey) _GATE(pkey)
#define DISABLE_PKRU     WRPKRU(0)

#endif

#define _GATE(pkey) GATE_##pkey

#define GATE_UNTRUSTED   WRPKRU(0xFFFFFFFC)
#define GATE_0           WRPKRU(0xFFFFFFF0)
#define GATE_1           WRPKRU(0xFFFFFFCC)
#define GATE_2           WRPKRU(0xFFFFFF3C)
#define GATE_3           WRPKRU(0xFFFFFCFC)
#define GATE_4           WRPKRU(0xFFFFF3FC)
#define GATE_5           WRPKRU(0xFFFFCFFC)
#define GATE_6           WRPKRU(0xFFFF3FFC)
#define GATE_7           WRPKRU(0xFFFCFFFC)
#define GATE_8           WRPKRU(0xFFF3FFFC)
#define GATE_9           WRPKRU(0xFFCFFFFC)
#define GATE_10          WRPKRU(0xFF3FFFFC)
#define GATE_11          WRPKRU(0xFCFFFFFC)
#define GATE_12          WRPKRU(0xF3FFFFFC)
#define GATE_13          WRPKRU(0xCFFFFFFC)
#define GATE_14          WRPKRU(0x3FFFFFFC)

// FIXME: This doesn't support arguments on the stack or in r9 (the last
// register). Also doesn't support returning 128-bit values (rdx).
#define INDIRECT_WRAPPER(target, caller_pkey, target_pkey)                     \
    /* Jump to a subsection of .text to prevent inlining this function */      \
    ".text 1\n"                                                                \
    /* This is the symbol name that will appear in the executable */           \
    "__ia2_" UNIQUE_STR(#target) "_wrapper:\n"                                 \
    /* Set the value of the wrapper name used by asm to this location */       \
    ".equ __ia2_" UNIQUE_STR(#target) ",.\n"                                   \
    DISABLE_PKRU                                                               \
    "movq " UNIQUE_STR(#target) "@GOTPCREL(%rip), %r9\n"                \
    "movq (%r9), %r9\n"                                                 \
    /* Switch PKRU to the target compartment */                                \
    GATE(target_pkey)                                                          \
    "callq *%r9\n"                                                                \
    /* Save return value before toggling PKRU */                               \
    "movq %rax, %r9\n"                                                            \
    GATE(caller_pkey)                                                          \
    /* Put return value back in the right register */                          \
    "movq %r9, %rax\n"                                                            \
    "ret\n"                                                                    \
    /* Jump back to the previous location in .text */                          \
    ".previous\n"                                                              \

// Takes a function pointer `target` and returns an opaque pointer for its call gate wrapper.
#define IA2_FNPTR_WRAPPER(target, ty, caller_pkey, target_pkey)    ({    \
    static char *target_ptr __asm__(UNIQUE_STR(#target));                \
    static void *wrapper __asm__("__ia2_" UNIQUE_STR(#target));          \
    target_ptr = (char *)target;                                         \
    __asm__(INDIRECT_WRAPPER(target, caller_pkey, target_pkey));         \
    (struct IA2_fnptr_##ty) {                                            \
        (char *)&wrapper                                                 \
    };                                                                   \
})

// Takes an opaque pointer `target` and returns a function pointer for its call gate wrapper.
#define IA2_FNPTR_UNWRAPPER(target, ty, caller_pkey, target_pkey)  ({    \
    static char *target_ptr __asm__(UNIQUE_STR(#target));                \
    static void *wrapper __asm__("__ia2_" UNIQUE_STR(#target));          \
    target_ptr = target.ptr;                                             \
    __asm__(INDIRECT_WRAPPER(target, caller_pkey, target_pkey));         \
    (IA2_FNPTR_TYPE_##ty)&wrapper;                                       \
})

// We must declare the sections used to pad the end of each program header
// segment to make sure their rwx permissions match the segment they're placed
// in. Otherwise if the padding sections are declared in the linker script
// without any input sections they and their corresponding segment will default
// to rwx. We avoid using .balign to align the sections at the start of each
// segment because it inserts a fill value (defaults to 0) which may break some
// sections (e.g.  insert null pointers into .init_array).
#define NEW_SECTION(name)          \
    __asm__(".section " #name "\n" \
            ".previous");

// Initializes a compartment with the nth key in IA2_INIT_DATA when the ELF
// invoking this is loaded. This must only be called once for each key. The
// compartment includes all segments in the ELF except the `ia2_shared_data`
// section, if one exists.
#define INIT_COMPARTMENT(n)                                          \
    NEW_SECTION(".fini_padding");                                    \
    NEW_SECTION(".rela.plt_padding");                                \
    NEW_SECTION(".eh_frame_padding");                                \
    NEW_SECTION(".bss_padding");                                     \
    struct IA2PhdrSearchArgs {                                       \
        int32_t pkey;                                                \
        void *address;                                               \
    };                                                               \
    __attribute__((constructor(102))) static void init_pkey_ctor() { \
        struct IA2PhdrSearchArgs args = {                            \
            .pkey = n + 1,                                           \
            .address = &init_pkey_ctor,                              \
        };                                                           \
        dl_iterate_phdr(protect_pages, &args);                       \
    }

#define INIT_RUNTIME(n_pkeys)                                                         \
    __attribute__((constructor(101))) static void init_runtime() {                    \
        for (int pkey = 0; pkey < n_pkeys; pkey++) {                                  \
            int allocated = pkey_alloc(0, 0);                                         \
            if (allocated != pkey + 1) {                                              \
                printf("Failed to allocate protection keys in the expected order\n"); \
                exit(0);                                                              \
            }                                                                         \
        }                                                                             \
    }
