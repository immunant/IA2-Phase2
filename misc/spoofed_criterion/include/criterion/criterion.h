#include <criterion/logging.h>
#include <criterion/new/assert.h>

#define Test(suite, name) \
    int fake_criterion_##suite##_##name(void); \
    __attribute__((__section__("fake_criterion_tests"))) \
        int (*fake_criterion_##suite##_##name##_##test)(void) = fake_criterion_##suite##_##name; \
    int fake_criterion_##suite##_##name(void)
