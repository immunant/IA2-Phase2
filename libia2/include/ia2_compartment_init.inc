/* -*-c-*- vi:filetype=c:
 */
#ifndef IA2_COMPARTMENT
#error Missing IA2_COMPARTMENT
#endif

#include <ia2.h>

#define CONCAT_(x, y) x##y
#define CONCAT(x, y) CONCAT_(x, y)
#define COMPARTMENT_IDENT(name) CONCAT(name##_, IA2_COMPARTMENT)

/* Fully expand x to a string if x is a macro (e.g. IA2_COMPARTMENT) */
#define XSTR(x) STR(x)
#define STR(x) #x

__thread void *COMPARTMENT_IDENT(ia2_stackptr) __attribute__((used));

/*
** Initializes a compartment with protection key `n`. Called automatically as a
** constructor when the ELF defining this symbol is loaded. This must only be
** called once for each key. The compartment includes all segments in the ELF
** except the `ia2_shared_data` section, if one exists.
**/

extern int ia2_n_pkeys_to_alloc;
void ensure_pkeys_allocated(int *n_to_alloc);
__attribute__((constructor)) static void COMPARTMENT_IDENT(init_pkey)() {
  ensure_pkeys_allocated(&ia2_n_pkeys_to_alloc);
  struct PhdrSearchArgs args = {
      .pkey = IA2_COMPARTMENT,
      .address = &COMPARTMENT_IDENT(init_pkey),
  };

  __asm__ volatile(
      /* Set PKRU to the compartment's value */
      "xorl %%ecx, %%ecx\n"
      "xorl %%edx, %%edx\n"
      "mov_pkru_eax " XSTR(IA2_COMPARTMENT) "\n"
      IA2_WRPKRU "\n"
      :
      :
      : "rax", "rcx", "rdx");
  dl_iterate_phdr(protect_pages, &args);
  __asm__ volatile(
      /* Set PKRU to fully untrusted (no access) */
      "xorl %%ecx, %%ecx\n"
      "xorl %%edx, %%edx\n"
      "xorl %%eax, %%eax\n"
      IA2_WRPKRU "\n"
      :
      :
      : "rax", "rcx", "rdx");
}

void COMPARTMENT_IDENT(init_tls)(void) {
  struct PhdrSearchArgs args = {
      .pkey = IA2_COMPARTMENT,
      .address = &COMPARTMENT_IDENT(init_tls),
  };
  dl_iterate_phdr(protect_tls_pages, &args);
}
