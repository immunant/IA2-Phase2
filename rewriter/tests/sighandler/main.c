/*
*/
#include "lib.h"
#include <stddef.h>
#include <signal.h>
#include <ia2.h>
#define IA2_DEFINE_TEST_HANDLER
#include <test_fault_handler.h>
#include <criterion/criterion.h>

INIT_RUNTIME(2);
#define IA2_COMPARTMENT 1
#include <ia2_compartment_init.inc>


struct handler {
    IA2_IGNORE(void (*handler)(int sig));
};

int main_secret = 3;

void trap_handler(int sig) {
    volatile int test = main_secret;
    volatile int test2 = CHECK_VIOLATION(lib_secret);
};

IA2_DEFINE_SIGHANDLER(trap_handler, 1);

void install_sighandler(struct handler *h) {
    static struct sigaction sa;
    memset(&sa, 0, sizeof(struct sigaction));
    sigemptyset(&(sa.sa_mask));

    if (h) {
        sa.sa_handler = h->handler;
    } else {
        sa.sa_handler = IA2_SIGHANDLER(trap_handler);
    }
    sigaction(SIGTRAP, &sa, NULL);
}

Test(sighandler, main_rewriter) {
    install_sighandler(NULL);
    raise(SIGTRAP);
}

Test(sighandler, main_ignore) {
    static struct handler h = {
        .handler = IA2_SIGHANDLER(trap_handler),
    };
    install_sighandler(&h);
    raise(SIGTRAP);
}

Test(sighandler, lib_rewriter) {
    install_sighandler_in_lib(true);
    test_handler_from_lib();
}

Test(sighandler, lib_ignore) {
    install_sighandler_in_lib(false);
    test_handler_from_lib();
}