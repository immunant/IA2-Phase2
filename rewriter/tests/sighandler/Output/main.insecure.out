trap_handler: main_secret is 3
trap_handler: lib_secret is CHECK_VIOLATION: did not seg fault as expected
